keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
++type RuntimeClassInfo struct {
++	// Runtime class name.
++	// Empty for the default runtime class.
++	// +optional
++	Name string
++	// RecursiveReadOnlyMounts is set to true if the runtime class supports RecursiveReadOnlyMounts.
++	// +optional
++	RecursiveReadOnlyMounts *bool
++}
++
+ // NodeSystemInfo is a set of ids/uuids to uniquely identify the node.
+ type NodeSystemInfo struct {
+ 	// MachineID reported by the node. For unique machine identification
+@@ -4472,6 +4528,9 @@ type NodeSystemInfo struct {
+ 	OperatingSystem string
+ 	// The Architecture reported by the node
+ 	Architecture string
++	// The available runtime classes.
++	// +optional
++	RuntimeClasses []RuntimeClassInfo
+ }
+ 
+ // NodeConfigStatus describes the status of the config assigned by Node.Spec.ConfigSource.
+```
+
+### CRI API
+
+Add `bool recursive_read_only` to the [`Mount`](https://github.com/kubernetes/cri-api/blob/v0.26.1/pkg/apis/runtime/v1/api.proto#L212-L224) message.
+CRI implementations will also expose the availability of the feature via the `RuntimeHandlerFeatures` message.
+
+As kubelet can inspect the availability of the feature via the `RuntimeHandlerFeatures` message,
+there is no concept of "IfPossible" in the CRI API;
+kubelet translates an "IfPossible" value in the Core API into true or false in the CRI API
+
+The `RuntimeHandlerFeatures` message is also propagated to the `NodeSystemInfo` struct of the Core API.
+
+Diff:
+```diff
+diff --git a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+index 11d58636369..3bbeeca4cb6 100644
+--- a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
++++ b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+@@ -217,6 +217,8 @@ message Mount {
+     // follow the symlink and mount the real destination to container.
+     string host_path = 2;
+     // If set, the mount is read-only.
++    // In the CRI API, readonly always defaults to false, regardless to the value of
++    // recursive_read_only.
+     bool readonly = 3;
+     // If set, the mount needs SELinux relabeling.
+     bool selinux_relabel = 4;
+@@ -226,6 +228,15 @@ message Mount {
+     repeated IDMapping uidMappings = 6;
+     // GidMappings specifies the runtime GID mappings for the mount.
+     repeated IDMapping gidMappings = 7;
++    // If set to true, the mount is made recursive read-only.
++    // In this CRI API, recursive_read_only is a plain true/false boolean, although its equivalent
++    // in the Kubernetes core API is a quaternary that can be nil, "Enabled", "IfPossible", or "Disabled".
++    // kubelet translates that quaternary value in the core API into a boolean in this CRI API.
++    // Remarks:
++    // - nil is just treated as false
++    // - when set to true, readonly must be explicitly set to true, and propagation must be PRIVATE (0).
++    // - (readonly == false && recursive_read_only == false) does not make the mount read-only.
```

Comment:
we need to declare behavior when the field was passed as not-`nil`, while the `RuntimeHandlerFeatures` was saying that this is not supported yet. On old runtimes it will be ignored. Should the runtimes which explicitly returned "not supported" also ignore it for consistency or should actively fail as an assert?